src/script/sign.h

```diff
@@ -32,6 +62,9 @@ class BaseSignatureCreator {
     /** Create a singular (non-script) signature. */
     virtual bool CreateSig(const SigningProvider& provider, std::vector<unsigned char>& vchSig, const CKeyID& keyid, const CScript& scriptCode, SigVersion sigversion) const =0;
     virtual bool CreateSchnorrSig(const SigningProvider& provider, std::vector<unsigned char>& sig, const XOnlyPubKey& pubkey, const uint256* leaf_hash, const uint256* merkle_root, SigVersion sigversion) const =0;
+
+    //! Fetch a preimage corresponding to this hash.
+    virtual bool LookupPreimage(SignatureData& sig_data, const std::vector<unsigned char>& hash, std::vector<unsigned char>& preimage) const =0;
```

Comment:
I find slightly ugly to expose `SignatureData` to the `SignatureCreator` interface. I'd say that `SignatureData` is the "upper" interface of the script/sign logic, which it exposes to its users. And `SignatureCreator` is part of the "lower" interface - it's how one can provide simple individual signature logic to script/sign. So it feels awkward to make the `SignatureCreator` implementations (which should deal with simple low-level things like signatures) be exposed to this `SignatureData` class which deals with script signing. This isn't more than a philosophical concern about code structure, but I'd like to offer an alternative.

Trying to draw an analogy with how keys/signatures are handled, there is the somewhat weird observation that hash preimages are both key material (which should really go in a `SigningProvider`, as it's a kind of secret) and signature material (which rightfully belongs in `SigningData`, as something that will end up in the witness). I don't think that's a problem - maybe it belongs in both. So a possibility is:
* Add functionality to `SigningProvider` to hold/lookup preimages.
* Have an overridable LookupPreimage in `SigningProvider` that takes a `const SigningProvider&` as argument (like CreateSig etc). It's so simple it could have a default implementation, looking things up in the provider. Dummy implementations could override it by return success with zero-byte ones.
* Still have preimage data in `SignatureData` too, accessed/controlled by a non-overridable ::CreatePreimage function (mimicking ::CreateSig in script/sign.cpp), which can invoke LookupPreimage on the `SignatureCreator` if needed, but can also access the "cached" version in SignatureData.

This means that `SignatureCreator` implementations only need to know about `SigningProvider`, like before, but can still override. It also avoids the need to pass a `SignatureData` to SignSignature.

---

An alternative, if the above is too complex, is to introduce a simple `PreimageData` structure, make `LookupPreimage` take `const PreimageData&`, and add a `PreimageData` field to `SignatureData`.

---

Lastly, if none of the above are doable: can `SignatureData&` here and the argument to SignSignature be made a const reference?

Reply:
I see. But since `SignatureData` is already largely exposed to the inner of `ProduceSignature`, what would you think of just making it a parameter in `SignSignature` (instead of creating a fresh one there)? Because actually it's not necessary to expose it to `SignatureCreator`:


<details>
<summary>Quick and dirty diff on the tip of the PR</summary>

```diff
diff --git a/src/script/sign.cpp b/src/script/sign.cpp
index 4fe738dfd5..f337e8ba48 100644
--- a/src/script/sign.cpp
+++ b/src/script/sign.cpp
@@ -89,13 +89,6 @@ bool MutableTransactionSignatureCreator::CreateSchnorrSig(const SigningProvider&
     return true;
 }
 
-bool MutableTransactionSignatureCreator::LookupPreimage(SignatureData& sig_data, const std::vector<unsigned char>& hash, std::vector<unsigned char>& preimage) const {
-    auto it = sig_data.hash_preimages.find(hash);
-    if (it == sig_data.hash_preimages.end()) return false;
-    preimage = it->second;
-    return true;
-}
-
 static bool GetCScript(const SigningProvider& provider, const SignatureData& sigdata, const CScriptID& scriptid, CScript& script)
 {
     if (provider.GetCScript(scriptid, script)) {
@@ -390,6 +383,17 @@ static CScript PushAll(const std::vector<valtype>& values)
     return result;
 }
 
+template<typename M, typename K, typename V>
+bool LookupHelper(const M& map, const K& key, V& value)
+{
+    auto it = map.find(key);
+    if (it != map.end()) {
+        value = it->second;
+        return true;
+    }
+    return false;
+}
+
 /**
  * Context for solving a Miniscript.
  * If enough material (access to keys, hash preimages, ..) is given, produces a valid satisfaction.
@@ -451,16 +455,16 @@ struct Satisfier {
 
     //! Hash preimage satisfactions.
     miniscript::Availability SatSHA256(const std::vector<unsigned char>& hash, std::vector<unsigned char>& preimage) const {
-        return m_creator.LookupPreimage(m_sig_data, hash, preimage) ? miniscript::Availability::YES : miniscript::Availability::NO;
+        return LookupHelper(m_sig_data.hash_preimages, hash, preimage) ? miniscript::Availability::YES : miniscript::Availability::NO;
     }
     miniscript::Availability SatRIPEMD160(const std::vector<unsigned char>& hash, std::vector<unsigned char>& preimage) const {
-        return m_creator.LookupPreimage(m_sig_data, hash, preimage) ? miniscript::Availability::YES : miniscript::Availability::NO;
+        return LookupHelper(m_sig_data.hash_preimages, hash, preimage) ? miniscript::Availability::YES : miniscript::Availability::NO;
     }
     miniscript::Availability SatHASH256(const std::vector<unsigned char>& hash, std::vector<unsigned char>& preimage) const {
-        return m_creator.LookupPreimage(m_sig_data, hash, preimage) ? miniscript::Availability::YES : miniscript::Availability::NO;
+        return LookupHelper(m_sig_data.hash_preimages, hash, preimage) ? miniscript::Availability::YES : miniscript::Availability::NO;
     }
     miniscript::Availability SatHASH160(const std::vector<unsigned char>& hash, std::vector<unsigned char>& preimage) const {
-        return m_creator.LookupPreimage(m_sig_data, hash, preimage) ? miniscript::Availability::YES : miniscript::Availability::NO;
+        return LookupHelper(m_sig_data.hash_preimages, hash, preimage) ? miniscript::Availability::YES : miniscript::Availability::NO;
     }
 };
 
@@ -720,11 +724,6 @@ public:
         sig.assign(64, '\000');
         return true;
     }
-
-    bool LookupPreimage(SignatureData& sig_data, const std::vector<unsigned char>& hash, std::vector<unsigned char>& preimage) const override {
-        preimage.assign(32, '\000');
-        return true;
-    }
 };
 
 }
diff --git a/src/script/sign.h b/src/script/sign.h
index fdb3244d5d..8df1fbc98d 100644
--- a/src/script/sign.h
+++ b/src/script/sign.h
@@ -62,9 +62,6 @@ public:
     /** Create a singular (non-script) signature. */
     virtual bool CreateSig(const SigningProvider& provider, std::vector<unsigned char>& vchSig, const CKeyID& keyid, const CScript& scriptCode, SigVersion sigversion) const =0;
     virtual bool CreateSchnorrSig(const SigningProvider& provider, std::vector<unsigned char>& sig, const XOnlyPubKey& pubkey, const uint256* leaf_hash, const uint256* merkle_root, SigVersion sigversion) const =0;
-
-    //! Fetch a preimage corresponding to this hash.
-    virtual bool LookupPreimage(SignatureData& sig_data, const std::vector<unsigned char>& hash, std::vector<unsigned char>& preimage) const =0;
 };
 
 /** A signature creator for transactions. */
@@ -83,7 +80,6 @@ public:
     const BaseSignatureChecker& Checker() const override { return checker; }
     bool CreateSig(const SigningProvider& provider, std::vector<unsigned char>& vchSig, const CKeyID& keyid, const CScript& scriptCode, SigVersion sigversion) const override;
     bool CreateSchnorrSig(const SigningProvider& provider, std::vector<unsigned char>& sig, const XOnlyPubKey& pubkey, const uint256* leaf_hash, const uint256* merkle_root, SigVersion sigversion) const override;
-    bool LookupPreimage(SignatureData& sig_data, const std::vector<unsigned char>& hash, std::vector<unsigned char>& preimage) const override;
 };
 
 /** A signature creator that just produces 71-byte empty signatures. */
```

</details>